bleak/backends/bluezdbus/manager.py

```diff
@@ -668,22 +693,66 @@ async def _wait_for_services_discovery(self, device_path: str) -> None:
 
         If a disconnect happens before the completion a BleakError exception is raised.
         """
-        services_discovered_wait_task = asyncio.create_task(
-            self._wait_condition(device_path, "ServicesResolved", True)
-        )
-        device_disconnected_wait_task = asyncio.create_task(
-            self._wait_condition(device_path, "Connected", False)
-        )
-        done, pending = await asyncio.wait(
-            {services_discovered_wait_task, device_disconnected_wait_task},
-            return_when=asyncio.FIRST_COMPLETED,
-        )
+        with contextlib.ExitStack() as stack:
+            services_discovered_wait_task = asyncio.create_task(
+                self._wait_condition(device_path, "ServicesResolved", True)
+            )
+            stack.callback(services_discovered_wait_task.cancel)
+
+            device_disconnected_wait_task = asyncio.create_task(
+                self._wait_condition(device_path, "Connected", False)
+            )
+            stack.callback(device_disconnected_wait_task.cancel)
 
-        for p in pending:
-            p.cancel()
+            # in some cases, we can get "InterfaceRemoved" without the
+            # "Connected" property changing, so we need to race against both
+            # conditions
+            device_removed_wait_task = asyncio.create_task(
+                self._wait_removed(device_path)
+            )
+            stack.callback(device_removed_wait_task.cancel)
+
+            done, _ = await asyncio.wait(
+                {
+                    services_discovered_wait_task,
+                    device_disconnected_wait_task,
+                    device_removed_wait_task,
+                },
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+
```

Comment:
Based on https://github.com/hbldh/bleak/pull/1329#discussion_r1273575569, it looks like we may need to check `done` for exceptions.

Reply:
Yes thats possible since tasks by default are not eager (at least by default and not unless https://github.com/python/cpython/issues/97696) and won't run until the next event loop so `_wait_condition` can try to access a removed key on the next iteration of the event loop